src/Riok.Mapperly/Descriptors/DescriptorBuilder.cs

```diff
@@ -106,24 +108,59 @@ private void ReserveMethodNames()
         }
     }
 
-    private void ExtractObjectFactories()
+    private IEnumerable<IUserMapping> ExtractUserMappings()
     {
-        _objectFactories = ObjectFactoryBuilder.ExtractObjectFactories(_builderContext, _mapperDescriptor.Symbol);
+        _mapperDescriptor.Static = _mapperDescriptor.Symbol.IsStatic;
+        IMethodSymbol? firstNonStaticUserMapping = null;
+
+        foreach (var userMapping in UserMethodMappingExtractor.ExtractUserMappings(_builderContext, _mapperDescriptor.Symbol))
+        {
+            // if a user defined mapping method is static, all of them need to be static to avoid confusion for mapping method resolution
+            // however, user implemented mapping methods are allowed to be static in a non-static context.
+            // Therefore we are only interested in partial method definitions here.
+            if (userMapping.Method is { IsStatic: true, IsPartialDefinition: true })
+            {
+                _mapperDescriptor.Static = true;
+            }
+            else if (firstNonStaticUserMapping == null && !userMapping.Method.IsStatic)
+            {
+                firstNonStaticUserMapping = userMapping.Method;
+            }
+
+            _mappings.Add(userMapping);
+
+            yield return userMapping;
+        }
+
+        if (_mapperDescriptor.Static && firstNonStaticUserMapping is not null)
+        {
+            _diagnostics.Add(
+                Diagnostic.Create(
+                    DiagnosticDescriptors.MixingStaticPartialWithInstanceMethod,
+                    firstNonStaticUserMapping.Locations.FirstOrDefault(),
+                    _mapperDescriptor.Symbol.ToDisplayString()
+                )
+            );
+        }
     }
 
-    private void ExtractUserMappings()
+    private ObjectFactoryCollection ExtractObjectFactories()
     {
-        foreach (var userMapping in UserMethodMappingExtractor.ExtractUserMappings(_builderContext, _mapperDescriptor.Symbol))
+        return ObjectFactoryBuilder.ExtractObjectFactories(_builderContext, _mapperDescriptor.Symbol);
+    }
+
+    private void EnqueueToBuildBody(IReadOnlyCollection<IUserMapping> userMappings, ObjectFactoryCollection objectFactoryCollection)
```

Comment:
I'd rename to `EnqueueUserMappings` (IMO that reflects better what happens here). Probably also simplify `objectFactoryCollection` to `objectFactories`.